### PR TITLE
Turn record access using string literals into static access operations

### LINKED ISSUE #564: Language Server: static access resolved but not found while hoering the element
**Describe the bug**

```nickel 
let x = { y = { z = true } }
in x."y".z
```

Neither `"y"` nor `z` can be used to find their declaration.

Reason for `"y"` : the corresponding linearization would list the `Usage` item and then the `Structure` item corresponding to the literal string "y".

### PR DESCRIPTION
**Pull request by ysndr**
Fixes #564

### CODE REVIEW

**Comment by ysndr** [comment]
The CI fail seems to be unrelated

```
thread 'record_contract' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass '--test examples'
```

**Comment by yannham** [comment]
Try to run the example using the `nickel` binary from your branch directly, like `nickel -f examples/whatever/foo.ncl export`. What happens sometime is that a failing test cause the `assert` panicking code to print huge terms, and the stack overflow arises during the printing, since test have a more limited stack. At least it happened before. On the other hand, the `nickel` command should just give you the output, or a clear error. (I'm not sure this is the case, but it's worth trying first, before more investigation)

**Comment by ysndr** [comment]
`for f in $(fd .ncl); ./target/debug/nickel -f $f export && echo`

Runs just fine,that is without stack overflows but lots of (probably expected) errors

<details>
<summary>
Output
</summary>

```
error: non serializable term
  ┌─ /Volumes/projects/nickel/benches/functions/church.ncl:2:12
  │
2 │   encode = fun n f x => if n == 0 then x else encode (n - 1) f (f x),
  │            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
   ┌─ /Volumes/projects/nickel/benches/lists/fold.ncl:24:13
   │
24 │         run = fun n =>
   │ ╭─────────────^
25 │ │         lists.foldl (fun x y => x*y + (x - y)) (lists.generate n (fun n => n/2))
   │ ╰────────────────────────────────────────────────────────────────────────────────^

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
  ┌─ /Volumes/projects/nickel/benches/lists/map.ncl:3:9
  │
3 │     run = fun n =>
  │ ╭─────────^
4 │ │     lists.map f (lists.generate functions.id n)
  │ ╰───────────────────────────────────────────────^

error: non serializable term
  ┌─ /Volumes/projects/nickel/benches/lists/pipe.ncl:2:9
  │
2 │     run = fun n =>
  │ ╭─────────^
3 │ │     let l = lists.generate (fun n => n*n) n in
4 │ │     let l = lists.filter (fun n => n % 2 == 0) l in
5 │ │     let l = lists.map (fun n => [n, n+1]) l in
6 │ │     let l = lists.flatten l in
7 │ │     lists.partition (fun n => n % 2 == 0) l
  │ ╰───────────────────────────────────────────^

error: non serializable term
  ┌─ /Volumes/projects/nickel/benches/numeric/fibonacci.ncl:3:11
  │
3 │         f = fun n =>
  │ ╭───────────^
4 │ │         if n == 0 || n == 1 then
5 │ │           1
6 │ │         else
7 │ │           f (n - 1) + f (n - 2)
  │ ╰───────────────────────────────^

error: non serializable term
   ┌─ /Volumes/projects/nickel/benches/numeric/pidigits.ncl:16:10
   │
16 │   atan = fun x iterCount => sum x 0 0 iterCount,
   │          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
  ┌─ /Volumes/projects/nickel/benches/numeric/reduce.ncl:3:11
  │
3 │       run = fun n =>
  │ ╭───────────^
4 │ │       let numbers = lists.generate functions.id n in
5 │ │       lists.foldl (fun acc x => acc + x) 0 numbers
  │ ╰──────────────────────────────────────────────────^

error: non serializable term
   ┌─ /Volumes/projects/nickel/benches/numeric/scalar.ncl:9:9
   │
 9 │     run = fun n =>
   │ ╭─────────^
10 │ │     let left = lists.generate functions.id n in
11 │ │     let right = lists.generate (fun n => n*n/2) n in
12 │ │     let prod = map2 (fun x y => x * y) left right in
13 │ │     lists.foldl (fun x y => x + y) 0 prod
   │ ╰─────────────────────────────────────────^

error: non serializable term
  ┌─ /Volumes/projects/nickel/benches/records/countLetters.ncl:2:9
  │
2 │     run = fun s =>
  │ ╭─────────^
3 │ │     let updateDict = fun dict char =>
4 │ │       if records.has_field char dict then
5 │ │         dict -$ char & {"#{char}" = dict."#{char}" + 1}
6 │ │       else
7 │ │         dict & {"#{char}" = 1} in
8 │ │     lists.foldl updateDict {} (strings.chars s)
  │ ╰───────────────────────────────────────────────^

error: non serializable term
   ┌─ /Volumes/projects/nickel/benches/records/merge.ncl:2:9
   │
 2 │     run = fun n m =>
   │ ╭─────────^
 3 │ │     let prefix = fun n =>
 4 │ │       let l = lists.generate (fun _n => "a") n in
 5 │ │       lists.foldl (fun x y => x ++ y) "" l in
   · │
17 │ │     let l = lists.map makeRec topList in
18 │ │     lists.foldl (fun r1 r2 => r1 & r2) {} l
   │ ╰───────────────────────────────────────────^

error: non serializable term
    ┌─ /Volumes/projects/nickel/benches/strings/interpolation.ncl:2:9
    │
  2 │     run = fun n => if n < 1 then
    │ ╭─────────^
  3 │ │     ""
  4 │ │       else m#"
  5 │ │         5bSVkTosPNAmX5yhmsDM
    · │
106 │ │         a9TLAq2eVr0exurkvJZN
107 │ │       "#m
    │ ╰─────────^

error: Blame error: contract broken by a value [extra fields `sum`,`jump`].
   ┌─ :1:1
   │
 1 │ #Checked
   │ -------- expected type
   │
   ┌─ /Volumes/projects/nickel/examples/a.ncl:25:5
   │
25 │   } = {
   │ ╭─────'
26 │ │   b."string" = true,
27 │ │   jump = b.string,
28 │ │   sum = func 1 1 1
29 │ │ }  in
   │ ╰─' evaluated to this expression
   │
   ┌─ <unknown> (generated by evaluation):1:1
   │
 1 │ <value=%923>
   │ ------------ evaluated to this value

note:
   ┌─ /Volumes/projects/nickel/examples/a.ncl:46:11
   │
46 │ (config | #Checked).b.string
   │           ^^^^^^^^ bound here

{
  "c": {
    "d": {
      "e": true
    }
  }
}
true
{
  "flags": [
    "Wextra",
    "ostuff.o"
  ],
  "optimization_level": 2,
  "path_libc": "/lib/x86_64-linux-gnu/libc.so"
}
55
5040
{
  "firewall": {
    "enabled": false,
    "open_ports": [
      21,
      80,
      443
    ],
    "type": "iptables"
  },
  "server": {
    "host": {
      "ip": "182.168.1.1",
      "name": "hello-world.net",
      "options": "TLS",
      "port": 80
    }
  }
}
{
  "firewall": {
    "enabled": true,
    "open_ports": [
      21,
      80,
      443
    ],
    "type": "iptables"
  },
  "server": {
    "host": {
      "options": "TLS"
    }
  }
}
{
  "server": {
    "host": {
      "ip": "182.168.1.1",
      "name": "hello-world.net",
      "port": 80
    }
  }
}
true
error: Blame error: contract broken by a value.
   ┌─ :1:1
   │
 1 │ List (#Container)
   │ ----------------- expected type
   │
   ┌─ /Volumes/projects/nickel/examples/record-contract/record-contract.ncl:40:16
   │
40 │     containers = {
   │ ╭────────────────'
41 │ │     "main container" = {
42 │ │       image = "k8s.gcr.io/#{name_}",
43 │ │       ports = [ 80 ]
44 │ │     }
45 │ │   }
   │ ╰───' evaluated to this expression
   │
   ┌─ <unknown> (generated by evaluation):1:1
   │
 1 │ { ... }
   │ ------- evaluated to this value

note:
   ┌─ /Volumes/projects/nickel/examples/record-contract/record-contract.ncl:26:16
   │
26 │   containers | List #Container,
   │                ^^^^^^^^^^^^^^^ bound here

false
error: Blame error: contract broken by a value.
   ┌─ :1:1
   │
 1 │ #DivBy3
   │ ------- expected type
   │
   ┌─ /Volumes/projects/nickel/examples/simple-contracts/simple-contract-div.ncl:15:2
   │
15 │ (4 | #Even
   │  - evaluated to this expression

note:
   ┌─ /Volumes/projects/nickel/examples/simple-contracts/simple-contract-div.ncl:16:6
   │
16 │    | #DivBy3)
   │      ^^^^^^^ bound here

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
   ┌─ /Volumes/projects/nickel/stdlib/contracts.ncl:45:19
   │
45 │     record_extend = fun field contr cont acc l t =>
   │ ╭───────────────────^
46 │ │       if %has_field% field t then
47 │ │           let acc = acc$[field = contr (%go_field% field l) (t."#{field}")] in
48 │ │           let t = t -$ field in
49 │ │           cont acc l t
50 │ │       else
51 │ │           %blame% (%tag% "missing field `#{field}`" l),
   │ ╰──────────────────────────────────────────────────────^

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non serializable term
   ┌─ <stdlib/contracts.ncl>:14:11
   │
14 │           (fun x => t (%go_codom% l) (e (s (%chng_pol% (%go_dom% l)) x)))
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

true
error: Unbound identifier
  ┌─ /Volumes/projects/nickel/tests/destructuring/assign_fail.ncl:3:1
  │
3 │ a + b == 3
  │ ^ this identifier is unbound

true
error: Blame error: contract broken by a value [extra field `b`].
  ┌─ :1:1
  │
1 │ Num
  │ --- expected type
  │
  ┌─ /Volumes/projects/nickel/tests/destructuring/closed_fail.ncl:2:11
  │
2 │ let {a} = {a=1, b=2} in
  │           ---------- evaluated to this expression
  │
  ┌─ <unknown> (generated by evaluation):1:1
  │
1 │ { ... }
  │ ------- evaluated to this value

note:
  ┌─ /Volumes/projects/nickel/tests/destructuring/closed_fail.ncl:1:1
  │
1 │ ╭
2 │ │ let {a} = {a=1, b=2} in
  │ ╰^ bound here

true
true
true
true
true
error: Missing field
  ┌─ /Volumes/projects/nickel/tests/destructuring/rest_fail.ncl:3:1
  │
3 │ y.a
  │ ^^^ this requires field a to exist

true
true
error: Blame error: contract broken by a value.
  ┌─ :1:1
  │
1 │ Str
  │ --- expected type
  │
  ┌─ /Volumes/projects/nickel/tests/destructuring/typecontract_fail.ncl:2:20
  │
2 │ let {a | Str} = {a=1} in
  │                    - evaluated to this expression

note:
  ┌─ /Volumes/projects/nickel/tests/destructuring/typecontract_fail.ncl:2:10
  │
2 │ let {a | Str} = {a=1} in
  │          ^^^ bound here

error: Blame error: contract broken by a value.
  ┌─ :1:1
  │
1 │ Str
  │ --- expected type
  │
  ┌─ /Volumes/projects/nickel/tests/imports/contract-fail.ncl:1:1
  │
1 │ 1 | Str
  │ - evaluated to this expression

note:
  ┌─ /Volumes/projects/nickel/tests/imports/contract-fail.ncl:1:5
  │
1 │ 1 | Str
  │     ^^^ bound here

{
  "a": 1,
  "b": 1
}
{
  "a": 1
}
5
3
{
  "foo": "ab"
}
42
44
44
2
error: Incompatible types
  ┌─ /Volumes/projects/nickel/tests/imports/typecheck-fail.ncl:1:1
  │
1 │ false : Num
  │ ^^^^^ this expression
  │
  = The type of the expression was expected to be `Num`
  = The type of the expression was inferred to be `Bool`
  = These types are not compatible

true
true
true
true
true
true
true
true
error: non serializable term
  ┌─ /Volumes/projects/nickel/tests/pass/imported.ncl:3:1
  │
3 │ fun x => x
  │ ^^^^^^^^^^

true
true
true
true
true
true
true
true
error: Incompatible types
   ┌─ /Volumes/projects/nickel/tests/pass/typechecking.ncl:93:4
   │
93 │   ({ "#{if true then "foo" else "bar"}" = 2, }."bla" : Num),
   │    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this expression
   │
   = The type of the expression was expected to be `{bla: Num | _a}`
   = The type of the expression was inferred to be `{_: _b}`
   = These types are not compatible

2
true
```

</details>

**Comment by yannham** [comment]
It's hard to tell, you have to target the exact example to see if something is going wrong. The point is that because the tests may do `assert_eq`, they may print things that the standard `nickel` doesn't in case the assert fail. If by bad luck they print a closure, they may try to print the environment, which may be big (or even recursive). Hence a stack overflow. So, if a test fail because it returns `false` instead of true, the nickel execution will just return `false`, but the test may overflow.

That being said I'm not sure this is the case with examples. But we are not looking for a stack overflow specifically, we are looking for the result of the `nickel` command on the test that is overflowing.

File: src/parser/utils.rs [hunk 1]
```diff
@@ -95,6 +95,36 @@ impl InfixOp {
     }
 }
 
+/// Turn record destructors using literal chunks only into static accesses
```
Comment (yannham) [inline src/parser/utils.rs:98]:
```suggestion
/// Turn dynamic accesses using literal chunks only into static accesses
```
While destructor is technically correct, I don't think we use it elsewhere in the codebase.

**Review by yannham — APPROVED** [review]
It's good to go once this failing test mystery have been solved.

**Comment by ysndr** [comment]
It's in the typechecking now and I can confirm it does relate to this change:

```
/Volumes/projects/nickel/tests/pass/typechecking.ncl:93:4
   │
93 │   ({ "#{if true then "foo" else "bar"}" = 2, }."bla" : Num),
   │    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this expression
   │
   = The type of the expression was expected to be `{bla: Num | _a}`
   = The type of the expression was inferred to be `{_: _b}`
   = These types are not compatible
```

Im just not sure yet where this comes from

**Comment by yannham** [comment]
It may be fixed, but this is not a simple one. Basically it requires to change the way typechecking is done. For now, you can:

1. Change the field back to something dynamic, artificially: `{...}."#{"bl" ++ "a"}"`. At least we would keep the old test semantics, rather than just remove it.
2. Open an issue for this

**Comment by yannham** [comment]
Sorry, this is not my morning. It is actually expected: the typechecker can't prove that the left hand side have a field named `foo`. So it is totally fine that this is rejected, in fact. You can forget the 2. part. Just change the test as prescribed in 1.

**Pull request merged**